### PR TITLE
test: cover <CategoryComboTableBody/> & cartesian

### PR DESCRIPTION
**Pull request by Mohammer5**
Closes [DHIS2-14386](https://dhis2.atlassian.net/browse/DHIS2-14386)
Closes [DHIS2-14387](https://dhis2.atlassian.net/browse/DHIS2-14387)

[DHIS2-14386]: https://dhis2.atlassian.net/browse/DHIS2-14386?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DHIS2-14387]: https://dhis2.atlassian.net/browse/DHIS2-14387?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/shared/utils.test.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { cartesian, filterObject } from './utils.js'
+
+describe('cartesian', () => {
+    it('should multiple two matrices', () => {
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/shared/utils.test.js:16]: I've added two edge cases

File: src/shared/utils.test.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { cartesian, filterObject } from './utils.js'
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/shared/utils.test.js:1]: Thanks! I removed the duplicate file

**Comment by Mohammer5** [comment]
Thanks @adeldhis2! I've fixed the things you found

**Review by Birkbjo — COMMENTED** [review]
Overall looks good, thanks a lot for doing this!
Just a few minor comments.

I do miss one thing though - and that is testing that the cells/inputs are rendered in the correct order. However this would basically test `getSortedCoCsByCatComboId`-selector, so maybe better to there?

Basically testing that each input`<input name=dataElementId.cocId`> is the same order as the given mock-data.

**Review by Birkbjo — COMMENTED** [review]
EDIT: Seems like GitHub double-posted this, sorry for that!

Overall looks good, thanks a lot for doing this!
Just a few minor comments.

I do miss one thing though - and that is testing that the cells/inputs are rendered in the correct order. However this would basically test `getSortedCoCsByCatComboId`-selector, so maybe better to there?

Basically testing that each input`<input name=dataElementId.cocId`> is the same order as the given mock-data.

File: src/data-workspace/category-combo-table-body/category-combo-table-body.test.js [hunk 3]
```diff
@@ -0,0 +1,770 @@
...
+        const inputRows = result.getAllByTestId('dhis2-dataentry-tableinputrow')
+        expect(inputRows.length).toBe(15)
+
+        inputRows.forEach((inputRow, index) => {
+            const dataElement = tableDataElements[index]
+            const headerCell = getByText(inputRow, dataElement.displayName, {
+                selector: '[data-test="dhis2-dataentryapp-dataelementcell"]',
+            })
+
+            const inputCells = getAllByTestId(
```
Comment (Birkbjo) [inline src/data-workspace/category-combo-table-body/category-combo-table-body.test.js:501]: Should we also check that these render actual input fields with `dataElementId.cocId` as the name?

File: src/data-workspace/category-combo-table-body/category-combo-table-body.test.js [hunk 4]
```diff
@@ -0,0 +1,770 @@
...
+    categories,
+    categoryOptions,
+    categoryCombos,
+    dataElements,
+}
+
+describe('<CategoryComboTableBody />', () => {
+    useMetadata.mockReturnValue({ data: metadata })
+
+    it('should render 15 input rows and 5 columns per input', () => {
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/data-workspace/category-combo-table-body/category-combo-table-body.test.js:457]: Yeah, I like it when the description explains exactly what I'm expecting. But I suppose you can also get that information when looking at the `expect` calls. I've copied all suggestions (incl. the ones below)

File: src/data-workspace/category-combo-table-body/category-combo-table-body.test.js (hunk 3 above)
Comment (Mohammer5) [reply src/data-workspace/category-combo-table-body/category-combo-table-body.test.js:501]: That seems a little tricky. In that case I think it'd be better to check that a `<DataEntryField />` has been rendered. Whether an input has the correct name should be a different test. In this component a field could be an input, several inputs (radio buttons), a button or a textbox. That makes it quite hard to test it properly

**Review by adeldhis2 — APPROVED** [review]
Nice! Thank you 🙏

File: src/data-workspace/category-combo-table-body/category-combo-table-body.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data-workspace/category-combo-table-body/category-combo-table-body.test.js]: Nice, thanks. I think it's nice that the test describe what the component/what you're testing is supposed to do. As it can give an overview of the functionality. Also what I've seen most libraries do.
